drivers: gpio stm32 driver also get the pin state

FRASTM opened the pull request:
In the stm32 gpio driver, the gpio_stm32_get_config function
also returns the pin state HIGH or LOW) to match gpio_pin_get_config

The gpio_stm32_get_config function is modified to include the pin state HIGH/LOW  to the flags variable
This variable is of type gpio_flags_t.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/48007

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: gpio stm32 driver also get the pin state

In the stm32 gpio driver, the gpio_stm32_get_config function
also returns the pin state HIGH or LOW) to match gpio_pin_get_config
This is valid with CONFIG_GPIO_GET_CONFIG.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: gpio: stm32 driver uses a structure to hold the pin configuration

Define a pin structure to carry the pin configuration elements
such as push/pull, opendrain, input/output, high/low.
This structure will give the gpio_flags_t to the gpio_stm32_get_config.
Only the stm32F1x serie does not provide the pin config get function.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Removing the CONFIG_GPIO_GET_CONFIG for the stm32F1 serie. The GPIO port structure of that serie has no pull up/down.
The pin_get_config() function will not give a correct config for the pin.
Actually the stm32CUbe functions  LL_GPIO_SetPinPull and LL_GPIO_GetPinPull  exist but are accessing the  (ODR)  Output register (instead of a Pull up/down since it does no exist !).

FRASTM commented:
Requires PR https://github.com/zephyrproject-rtos/zephyr/pull/47923 for check PASSED with twister tests/drivers/gpio/gpio_basic_api on nucleo_f103rb

FRASTM commented:
rebase on 5897de1e535816247e309348cbe29760361a28b8

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -579,12 +585,13 @@ static int gpio_stm32_get_config(const struct device *dev,
 	otype = LL_GPIO_GetPinOutputType(gpio, pin_ll);
 	pupd = LL_GPIO_GetPinPull(gpio, pin_ll);
 	mode = LL_GPIO_GetPinMode(gpio, pin_ll);
+	set = LL_GPIO_IsOutputPinSet(gpio, pin_ll);
```
Pin could be "set" on IDR as well. Please use something more explicit, like 'out_set', 'oset', 'output_set', ...

FRASTM replied:
using a structure instead

erwango commented on drivers/gpio/gpio_stm32.h:
```diff
@@ -206,6 +206,18 @@
+	unsigned int type; /* LL_GPIO_OUTPUT_PUSHPULL or LL_GPIO_OUTPUT_OPENDRAIN */
+	unsigned int pupd; /* LL_GPIO_PULL_NO or LL_GPIO_PULL_UP or LL_GPIO_PULL_DOWN */
+	unsigned int mode; /* LL_GPIO_MODE_INPUT or LL_GPIO_MODE_OUTPUT or other */
+	unsigned int state; /* 1 (high level) or 0 (low level) */
```
```suggestion
	unsigned int out_state; /* 1 (high level) or 0 (low level) */
```
?

FRASTM replied:
change the name of the element in the `gpio_stm32_pin` structure

FRASTM commented on tests/drivers/gpio/gpio_basic_api/boards/nucleo_f103rb.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_GPIO_GET_CONFIG=n
```
The stm32f1 soc cannot force the CONFIG_GPIO_GET_CONFIG to 'n' through a kconfig
Could we then have  
```
#if defined( CONFIG_GPIO_GET_CONFIG) && defined(CONFIG_SOC_SERIES_STM32F1X)
#error stm32F1x does not support pin_get_config()
#endif
```

erwango replied:
> The stm32f1 soc cannot force the CONFIG_GPIO_GET_CONFIG to 'n' through a kconfig Could we then have
> 
> ```
> #if defined( CONFIG_GPIO_GET_CONFIG) && defined(CONFIG_SOC_SERIES_STM32F1X)
> #error stm32F1x does not support pin_get_config()
> #endif
> ```

Generating an error is not required, see API implementation. Not supporting this API call is legit:
https://github.com/zephyrproject-rtos/zephyr/blob/352fde1271c568fa24cb48b4aed20f97b25945b1/include/zephyr/drivers/gpio.h#L881-L883
If not implemented, caller will get `-ENOSYS` as a result. This is ok.

FRASTM committed:
tests: drivers: gpio basic pass testcase when test pin_get is not defined

When the CONFIG_GPIO_GET_CONFIG is set, even if the driver does not
provide the pin_get_config function, it is supposed to return -ENOSYS
error code (see zephyr/include/zephyr/drivers/gpio.h)
The testcase should accept this code and PASS.
(ENOTSUP is not the right value to check).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
add a flag CONFIG_SOC_SERIES_STM32F1X to avoid defining the pin_get_config function 
change the tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c  to handle the gpio_pin_get_config correctly

erwango commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -657,7 +657,7 @@ static int gpio_stm32_manage_callback(const struct device *dev,
-#ifdef CONFIG_GPIO_GET_CONFIG
+#if defined(CONFIG_GPIO_GET_CONFIG) && !defined(CONFIG_SOC_SERIES_STM32F1X)
 	.pin_get_config = gpio_stm32_get_config,
 #endif /* CONFIG_GPIO_GET_CONFIG */
```
This would be better in a dedicated commit

FRASTM committed:
drivers: gpio: stm32F1x serie does not provide the pin get config

The stm32F1x serie does not provide the pin_get_config function,
Even if the CONFIG_GPIO_GET_CONFIG is set, the -ENOSYS error code
is returned by the z_impl_gpio_pin_get_config().

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding a commit for stm32F1 serie : 168fbc8a22e38ebc7029277061eeb1cde5a8c24a

cfriedt commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -644,7 +644,7 @@ static int pin_get_config(void)
 
 	rc = gpio_pin_get_config(dev, PIN_OUT, &flags_get);
-	if (rc == -ENOTSUP) {
+	if (rc == -ENOSYS) {
```
I believe this should not be changed.

E.g. see
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/gpio/gpio_get_direction/src/main.c#L43

erwango replied:
> I believe this should not be changed.
> 
> E.g. see https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/gpio/gpio_get_direction/src/main.c#L43

@cfriedt, it seems the line you're pointing is not in line with gpio header:
https://github.com/zephyrproject-rtos/zephyr/blob/6cfb18686e1c494d0011aec87bbf24ab530d3a34/include/zephyr/drivers/gpio.h#L881-L882

cfriedt replied:
Oh weird - OK - can you make an additional commit to fix the test file that I pointed to? Thanks for pointing that out.
